converters/toZigbee.js

```diff
@@ -841,177 +842,69 @@ const converters = {
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
-                }
-            } else if (value.hasOwnProperty('hsv')) {
-                let hsv = value.hsv.split(',').map((i) => parseInt(i));
-                newState.color = {h: hsv[0], s: hsv[1], v: hsv[2]};
-                hsv = utils.gammaCorrectHSV(utils.correctHue(hsv[0], meta), hsv[1], hsv[2]);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
-                }
-            } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s')) {
-                newState.color = {h: value.h, s: value.s};
-                const hsv = utils.gammaCorrectHSV(utils.correctHue(value.h, meta), value.s, 100);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturation';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturation';
+            // TODO: define and implement convention what should be stored to state (ex. hsl vs {h,s,v})
+
+            if (newColor.xy !== null) {
+                value.x = newColor.xy.x;
+                value.y = newColor.xy.y;
+                newState.color = newColor.xy.toObject(4);
+            } else if (newColor.hsv !== null) {
+                const enhancedHue = utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true);
+                const hsv = newColor.hsv;
+                const hsvCorrected = hsv.colorCorrected(meta);
+                newState.color = hsv.toObject();
```

Comment:
should we store corrected or uncorrect hsv here?

Reply:
I thought about this and I think we should store uncorrected one.

This is because end user shouldn't be aware that any correction happens at all. He selects preferred color on the color wheel and expects same color from the controlled light. Color correction should map color wheel in the app to virtual, malformed color wheel of the device and result of this correction shouldn't be of interest of end user.

If we would store state after correction, position on color when would magically jump to some other coordinates, this would be counter intuitive IMHO.

Reply:
True, but the question is then... should we reverse the correction (can we?) when the device returns the value via a read/report ?

Reply:
That is the tricky part. In Tuya I'm just ignoring color reported by device and just go with optimistic updates.

@Koenkk what do you think about this issue?

Reply:
That should certainly be the default I think, see  https://github.com/Koenkk/zigbee-herdsman-converters/pull/2453#discussion_r607667272

Reply:
Reverse correction should be possible, but I think a better option would be to store the raw input in some hidden state.